### PR TITLE
feat(optimization): CPU cache control to enable hashrate optimizations

### PR DESCRIPTION
**Pull request by raftedproc**
This patch adds MSR optimization functionality into CCP.

### CODE REVIEW

**Comment by monoid** [comment]
Cargo.lock is not updated.

File: crates/msr/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "ccp-msr"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (mikevoronov) [inline crates/msr/Cargo.toml:6]: obsolete

File: crates/msr/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+[dependencies]
+cpu-utils = {workspace = true}
+thiserror = { workspace = true }
+nix = { version = "0.27.1" , features = ["uio"] }
```
Comment (mikevoronov) [inline crates/msr/Cargo.toml:11]: Let's specify versions in the workspace root and then just use { workspace = true} here.

File: crates/msr/src/msr.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+use std::fs::{File, OpenOptions};
+use std::io;
+
+use crate::cpu_preset::get_cpu_preset;
+use crate::msr_item::MsrItem;
+use crate::msr_mode::MSR_MODE;
+use crate::MSRError;
+use cpu_utils::LogicalCoreId;
+
+enum MsrFileOpMode {
```
Comment (mikevoronov) [inline crates/msr/src/msr.rs:28]: Let's use capital `MSR` everywhere for consistency

File: crates/msr/src/msr_error.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+use nix::errno::Errno;
+use thiserror::Error as ThisError;
+
+use cpu_utils::LogicalCoreId;
+
+#[derive(ThisError, Debug)]
+pub enum MSRError {
+    #[error("opening MSR file '/dev/cpu/{0:}/msr' for read there is an error: {1:?}")]
+    OpenForRead(LogicalCoreId, std::io::Error),
```
Comment (mikevoronov) [inline crates/msr/src/msr_error.rs:25]: Let's name fields in enum variants when there are more than 1 of them.

File: crates/msr/src/msr_error.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (mikevoronov) [inline crates/msr/src/msr_error.rs:1]: Let's name this file just `errors` for correspondence with other parts

**Review by mikevoronov — APPROVED** [review]
utACK

File: main/src/main.rs [hunk 6]
```diff
@@ -65,6 +66,9 @@ struct ProverArgs {
 
     #[arg(long)]
     dir_to_store_persistent_state: PathBuf,
+
+    #[arg(long, action = ArgAction::SetTrue)]
```
[the remark this answers is not in the document]
Comment (raftedproc) [reply main/src/main.rs:70]: It is already disabled by default. ArgAction sets the value when the flag is present.

**Review by mikevoronov — APPROVED** [review]

**Pull request merged**